libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/analysis/FunctionCallGraph.h>
+#include <boost/range/adaptor/reversed.hpp>
+
+using namespace std;
+using namespace solidity::frontend;
+
+FunctionCallGraphBuilder::FunctionCallGraphBuilder(ContractDefinition const& _contract):
+	m_contract(&_contract),
+	m_graph(std::make_unique<ContractCallGraph>(_contract))
+{
+	// Create graph for constructor, state vars, etc
+	m_currentNode = SpecialNode::EntryCreation;
+	m_currentDispatch = SpecialNode::InternalCreationDispatch;
+
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts | boost::adaptors::reversed)
+	{
+		for (auto const* stateVar: contract->stateVariables())
+			stateVar->accept(*this);
+
+		for (auto arg: contract->baseContracts())
+			arg->accept(*this);
+
+		if (contract->constructor())
+		{
+			add(*m_currentNode, contract->constructor());
+			contract->constructor()->accept(*this);
+			m_currentNode = contract->constructor();
+		}
+	}
+
+	m_currentNode.reset();
+	m_currentDispatch = SpecialNode::InternalDispatch;
+
+	// Create graph for all publicly reachable functions
+	for (auto& [hash, functionType]: _contract.interfaceFunctionList())
+	{
+		(void)hash;
+		if (auto const* funcDef = dynamic_cast<FunctionDefinition const*>(&functionType->declaration()))
+			if (!m_graph->edges.count(funcDef))
```

Comment:
I think we at least have to re-visit everything between creation and deployed analysis. The reason is that if you have a function `f` that calls a function `g` indirectly, the edge goes to the creation dispatch at creation time but to the deployed dispatch after deployment.

Reply:
(should go - currently it always goes to the creation dispatch)

Reply:
Yeah. And currently it goes to both dispatches in some cases (but not in others). See for example my `indirect_calls` test. The check against not visiting a function twice does not work for functions that make no calls so they may get visited both from the constructor and from a call to deployed code.

Splitting it into two separate graphs should solve that problem.